src/App.tsx

```diff
@@ -1,26 +1,304 @@
+        if (inputRef.current) {
+          inputRef.current.disabled = false;
+          inputRef.current.focus();
+        }
+
+        setTempTodo(null);
+        setLoadingTodos(current => current.filter(todoId => todoId !== 0));
+      });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setLoadingTodos(current => [...current, todoId]);
+
+    deleteTodo(todoId)
+      .then(() =>
+        setTodos(currentTodos =>
+          currentTodos.filter(todo => todo.id !== todoId),
+        ),
+      )
+      .catch(() => setErrorMessage('Unable to delete a todo'))
+      .finally(() => {
+        if (inputRef.current) {
+          inputRef.current.disabled = false;
+          inputRef.current.focus();
+        }
+
+        setLoadingTodos(current =>
+          current.filter(delitingTodoId => todoId !== delitingTodoId),
+        );
+      });
+  };
+
+  const handleDeleteAllCompleted = () => {
+    completedTodos.forEach(todo => handleDeleteTodo(todo.id));
+  };
+
+  const handleToggleStatus = (todo: Todo) => {
+    setLoadingTodos(current => [...current, todo.id]);
+
+    updateTodo({ ...todo, completed: !todo.completed })
+      .then(updatedTodo =>
+        setTodos(currentTodos =>
+          currentTodos.map(currentTodo =>
+            currentTodo.id === updatedTodo.id ? updatedTodo : currentTodo,
+          ),
+        ),
+      )
+      .catch(() => setErrorMessage('Unable to update a todo'))
+      .finally(() => {
+        if (inputRef.current) {
+          inputRef.current.disabled = false;
+          inputRef.current.focus();
+        }
+
+        setLoadingTodos(current => current.filter(id => id !== todo.id));
+      });
+  };
+
+  const handleAllToggleStatus = () => {
+    if (activeTodos.length > 0) {
+      activeTodos.forEach(handleToggleStatus);
+    } else {
+      completedTodos.forEach(handleToggleStatus);
+    }
+  };
+
+  const handleRename = (todo: Todo) => {
+    setLoadingTodos(current => [...current, todo.id]);
+
+    updateTodo(todo)
+      .then(updatedTodo =>
+        setTodos(currentTodos =>
+          currentTodos.map(currentTodo =>
+            currentTodo.id === updatedTodo.id ? updatedTodo : currentTodo,
+          ),
+        ),
+      )
+      .catch(() => setErrorMessage('Unable to update a todo'))
+      .finally(() => {
+        if (inputRef.current) {
+          inputRef.current.disabled = false;
+          inputRef.current.focus();
+        }
+
+        setLoadingTodos(current => current.filter(id => id !== todo.id));
+      });
+  };
+
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">
-          React Todo App - Add and Delete
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
```

Comment:
this component became quite big now
let's split it into a few separate components